Roll out new version of the project template (2018/12/15)

claudiofantacci opened the pull request:
With this PR we now have a refreshed, shiny and awesome CMake template for exporting Cpp libraries!

In particular I tried to address all the entries of #33 by
 - Migrating Travis to Xenial and Xcode9.4, using addons for Homebrew (see #29).
 - Create a matching alias library with namspace prefix, as reported in Chapter 16.2 of C. Scott's Professional CMake (see #30).
 - Add header-only library template (see #31). This change had quite some implications. I found it useful to have a template just for that, as emerged from #28. This also impied the need for the project to export multiple libraries, not just one. Consequently, I had to add a variable with parent scopes to backprop to the main CMakeLists the name of the targets.
 - Polish up ✨ the README 😎 (see #35).
 - Change PATH (legacy keyword) in DIRECTORY in `dox` custom target (see #38).
 - Last, but not least, I tried to provide guidelines and useful links about the complexity of mantaining a C++ library (see #34).

Let me know what you think and should you have any suggestion for improvements I'll try to address them asap! 👍 

Closes #29, closes #30, closes #31, closes #32, closes #34, closes #35, closes #33, closes #38.

claudiofantacci committed:
Update Travis script

claudiofantacci committed:
Use 2 spaces indentation in CI scripts

claudiofantacci committed:
Create matching alias with namespace for the library

claudiofantacci committed:
Add a second template for a header-only library

claudiofantacci committed:
Add a test for the header only library

claudiofantacci committed:
Clean up code

claudiofantacci committed:
Update the README

claudiofantacci committed:
Add guidelines for library maintenence

claudiofantacci committed:
Add new executable using header-only library

claudiofantacci committed:
Update tests

traversaro commented on CMakeLists.txt:
```diff
@@ -121,6 +121,11 @@ endif()
+# This variable will be filled by the CMakeLists of each library with their
+# target name. The target list will be used to create and install CMake
+# configuration files.
+set(${PROJECT_NAME}_TARGETS)
```
I am not a big fan of this "manual" handling of variables and manually dealing with stuff such as `PARENT_SCOPE` , but as far as I understand the alternative is just use a GLOBAL PROPERTY as done in iDynTree, that has the advantages of just requiring a one-iner to add a new variable to the list, see for example https://github.com/robotology/idyntree/blob/2597eac58f87446fd1f6261317a37fd46916a4ad/src/core/CMakeLists.txt#L142, but it no beatiful/superclear either. @drdanz what do you think of this?

claudiofantacci replied:
@traversaro I totally agree with you. As a matter of the fact, I tried to do just like that in the first place 😲, but It wasn't working properly 🙃 So I decided to go with `PARENT_SCOPE`d variable that I agree are super-duper-ugly to have and handle. @drdanz any suggestion? What is the cleanest approach to this?

claudiofantacci replied:
In particular I was not able to pass the targets up in the scope using the `set_property`command and, consequently, [`install_basic_package_files`](https://github.com/robotology/idyntree/blob/2597eac58f87446fd1f6261317a37fd46916a4ad/CMakeLists.txt#L67-L71) was not working properly.

claudiofantacci replied:
I addressed this using a `GLOBAL property`. See new https://github.com/robotology/how-to-export-cpp-library/pull/36/commits/479198992518cc52634fa095dcf717ec6c22267b commit.

traversaro commented on README.md:
```diff
@@ -8,9 +7,38 @@ This repository provides an OS-agnostic C++ library template with plain [CMake](
+ - [Good practices using Git branching](https://nvie.com/posts/a-successful-git-branching-model/)
+ - [Version handling](https://semver.org/)
+ - [Dynamic loading](https://amir.rachum.com/blog/2016/09/17/shared-libraries/)
+ - [API/ABI backward compatibility](https://abi-laboratory.pro/index.php?view=binary-compatibility)
```
I think two additional nice points to add would be: 
* [Symbol Visibility](https://www.akkadia.org/drepper/dsohowto.pdf)
* Not sure how, but I think https://github.com/green7ea/cpp-compilation/blob/master/README.md is a super good resource for people that are starting to seriously work with C/C++.

claudiofantacci replied:
I was unsure to add `How To Write Shared Libraries` because it was just too much, but since you are also suggesting it, then I'll just add it 😄 
The other reference is terrific! Added it right away! 👍

traversaro commented on .appveyor.yml:
```diff
@@ -3,28 +3,29 @@ version: 1.0.{build}
-    - Visual Studio 2017
+  - Visual Studio 2013
+  - Visual Studio 2015
+  - Visual Studio 2017
```
Not directly related to the PR, but I just noticed that on VS we just test using the 32-bit platform, never compiling for 64-bit .

claudiofantacci replied:
🙈 
I'll add the tests right away, deprecating `Visual Studio 2013`.

traversaro reviewed (dismissed):
Some minor comments, but they are not blocking the merge at all.
Thanks a lot @claudiofantacci this is amazing!

traversaro commented:
By the way, I noticed that we did not mention how to handle static variables in shared libraries in Windows. Unfortunately those are not handled by [`CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS`](https://blog.kitware.com/create-dlls-on-windows-without-declspec-using-new-cmake-export-all-feature/). While the blog post that we link in the README mention that, I wonder if we should stress this more, or we can simply ignore this relatively niche use case. See https://github.com/ms-iot/ROSOnWindows/issues/33 and  https://github.com/ms-iot/ROSOnWindows/pull/37 for a related discussion.

traversaro commented on README.md:
```diff
@@ -95,5 +133,15 @@ Should you not be sure what to do about it (licensing produces severe headhaces)
+
+
+---
+If you feel this CMake project template was useful, consider leaving the following shield somewhere in your README project (feel free to modify its look as you please). Otherwise, not a big deal! 👍
```
If we are interested in dissemination, we can also ask for a GitHub star.

claudiofantacci replied:
Good point 👍 What do you suggest to do? Remove this comment and kindly ask for stars, or to also add star?

claudiofantacci committed:
Update AppVeyor script

claudiofantacci commented:
> By the way, I noticed that we did not mention how to handle static variables in shared libraries in Windows. Unfortunately those are not handled by [`CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS`](https://blog.kitware.com/create-dlls-on-windows-without-declspec-using-new-cmake-export-all-feature/). While the blog post that we link in the README mention that, I wonder if we should stress this more, or we can simply ignore this relatively niche use case. See [ms-iot/ROSOnWindows#33](https://github.com/ms-iot/ROSOnWindows/issues/33) and [ms-iot/ROSOnWindows#37](https://github.com/ms-iot/ROSOnWindows/pull/37) for a related discussion.

Well I guess we can address this as well in this PR. I was unaware of this "obscure issues" (cit.) under Windows. I'll try to understand the content of the linked material (thanks about that!) and then I'll add one more commit to the PR.

In the meanwhile let's try to close all the ongoing discussions 👍

claudiofantacci commented:
> By the way, I noticed that we did not mention how to handle static variables in shared libraries in Windows. Unfortunately those are not handled by [`CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS`](https://blog.kitware.com/create-dlls-on-windows-without-declspec-using-new-cmake-export-all-feature/). While the blog post that we link in the README mention that, I wonder if we should stress this more, or we can simply ignore this relatively niche use case. See [ms-iot/ROSOnWindows#33](https://github.com/ms-iot/ROSOnWindows/issues/33) and [ms-iot/ROSOnWindows#37](https://github.com/ms-iot/ROSOnWindows/pull/37) for a related discussion.

Let me add one more reply about this suggestion.
I think we can give proper room to this particular use case, but it requires to change the library a little bit and probably add some more test in Appveyor.
As a matter of the fact, I would probably add a third library that would highlight this particular use case, but in another PR in the next future. What do you think about it?
We can convert your https://github.com/robotology/how-to-export-cpp-library/pull/36#issuecomment-447587556 into an issue for keeping the record of it 👍

claudiofantacci committed:
Clean up shared lib header code

claudiofantacci committed:
Change PATH with DIRECTORY in doxygen CMakeLists

claudiofantacci committed:
Clean up code style in dox CMakeLists

claudiofantacci committed:
Update Doxygen documentation

claudiofantacci commented:
Ping @drdanz 😁

drdanz commented on CMakeLists.txt:
```diff
@@ -119,8 +119,26 @@ if(NOT CMAKE_CONFIGURATION_TYPES)
+# with the set_property() and get_property() commands.
+# This is also useful to associate documentation with property names that may be
+# retrieved with the get_property() command in subdirectories.
+define_property(GLOBAL
```
:-1: -1 for using properties to remember the targets to pass to `install_basic_package_files`...
As already discussed in person, using `EXPORT` is the right way to do it, and `TARGETS` unofficially deprecated, since it requires a lot more work and could lead to differences between the build tree and the install tree

claudiofantacci replied:
@drdanz ok and thanks for the explanation. I'll update the PR asap and ping you back as soon as it is ready 👍

claudiofantacci commented:
I updated the PR following @drdanz suggestions.

I also added a long comment about the `FIRST_TARGET` option of `install_basic_package_files` [here](https://github.com/robotology/how-to-export-cpp-library/compare/80805e8ff05830aaac717dc527dfe7d0ec2165f4..cb349e486924b9fcdf4148d742d75f9d60690c97#diff-6853026e40dc273b21195c7c75fe7e19).
Please have a look at it and, if useful, we can polish and modify to including it as a comment of the option in `install_basic_package_files` itself.

drdanz reviewed (approved):
LGTM now!

claudiofantacci commented:
Great! Let's merge it! 🎉
